src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderPort.java

```diff
@@ -155,74 +155,57 @@ else if (paymentServiceProviderResult == PaymentServiceProviderResult.IDENTIFY_S
         return purchaseResult;
     }
 
-    public PurchaseResult authoriseKlarnaPayment(final String merchantAccount,
+    public PurchaseResult authoriseKlarnaPayment(final boolean authComplete,
+                                                 final String merchantAccount,
                                                  final PaymentData paymentData,
                                                  final UserData userData) {
-        PaymentsRequest klarnaRequest = adyenRequestFactory.createKlarnaPayment(merchantAccount, paymentData, userData);
-        PurchaseResult purchaseResult = null;
-        try {
-            PaymentsResponse response = adyenCheckoutApiClient.createPayment(klarnaRequest);
-            purchaseResult = extractKlarnaResponse(response);
+        final AdyenCallResult<PaymentsResponse> adyenCallResult;
+        if(authComplete) {
+            //completing auth session
+            PaymentsDetailsRequest detailsRequest = adyenRequestFactory.completeKlarnaPayment(merchantAccount, paymentData, userData);
+            adyenCallResult = adyenCheckoutApiClient.paymentDetails(detailsRequest);
+        } else {
+            //starting auth session
+            PaymentsRequest klarnaRequest = adyenRequestFactory.createKlarnaPayment(merchantAccount, paymentData, userData);
+            adyenCallResult = adyenCheckoutApiClient.createPayment(klarnaRequest);
         }
-        catch (ApiException ex) {
-            return new PurchaseResult(PaymentServiceProviderResult.ERROR,
-                    null,
-                    null,
-                    ex.getMessage(),
-                    null,
-                    klarnaRequest.getReference(), null);
-        } catch (IOException ex) {
-            return new PurchaseResult(PaymentServiceProviderResult.ERROR,
-                    null,
-                    null,
-                    ex.getMessage(),
-                    null,
-                    klarnaRequest.getReference(), null);
+
+        if (!adyenCallResult.receivedWellFormedResponse()) {
+            return handleKlarnaAuthoriseError(adyenCallResult, paymentData);
         }
 
-        return purchaseResult;
+        return handleKlarnaAuthoriseResponse(adyenCallResult, merchantAccount);
```

Comment:
I think checkout API is generic. We should have generic Response handling. 
Do we really need Klarna specific error and response handling ?